google/cloud/spanner_v1/merge_result_set.py

```diff
@@ -0,0 +1,114 @@
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from queue import Queue
+from typing import Any, TYPE_CHECKING
+from threading import Lock
+
+if TYPE_CHECKING:
+    from google.cloud.spanner_v1.database import BatchSnapshot
+
+QUEUE_SIZE_PER_WORKER = 32
+LOCK = Lock()
+
+
+def _set_metadata(merged_result_set, results):
+    LOCK.acquire()
+    try:
+        merged_result_set._metadata = results.metadata
+        merged_result_set._stats = results.stats
+        merged_result_set._first_row_read = True
+    finally:
+        LOCK.release()
+
+
+class PartitionExecutor:
+    """
+    Executor that executes single partition on a separate thread and inserts
+    rows in the queue
+    """
+
+    def __init__(self, batch_snapshot, partition_id, merged_result_set):
+        self._batch_snapshot: BatchSnapshot = batch_snapshot
+        self._partition_id = partition_id
+        self._merged_result_set: MergedResultSet = merged_result_set
+        self._queue: Queue[PartitionExecutorResult] = merged_result_set._queue
+
+    def run(self):
+        try:
+            results = self._batch_snapshot.process_query_batch(self._partition_id)
+            merged_result_set = self._merged_result_set
+            for row in results:
+                if not merged_result_set._first_row_read:
```

Comment:
As all partitions would have same metadata so any of the partition can set it. Also if 2 threads see its value as False because of syncronization issues then also it should work fine as 2nd then thread would just override with same value. 
As I am thinking about it I guess even we dont need lock for writing as well as the value of `_first_row_read` only changes from False to True in its lifecycle and other threads would simply set the value to True again

Reply:
Well, it depends on how Python handles this. If you were to do it like this in:
1. Java: It would work, but it could be that `_first_row_read` would be set to true multiple times (so not really a problem)
2. Go: It would crash the application because of a data race (or `panic` as Go so nicely calls it). See https://go.dev/doc/articles/race_detector

So we should make sure that Python does not behave the same as Go would in this case.

Reply:
Will find out what happens in python but we can continue using the lock for writing (as already done in this PR) to be safe

Reply:
Removed the _first_row_read property and checking if _metadata is None